### ISSUE TITLE
Offer DB backend choice between SQLite and xSQL

### ISSUE THREAD

**Issue by DragonSupremacy** [StrangeLoopGames/EcoSuggestions#651]
Being an admin/server owner for about 2 months now, I'm still amazed at how quickly the SQLite .db file balloons in size even with just several players playing simultaneously. I can only imagine what sort of performance impact this has on larger Eco servers. 

Over the course of just 4 days with anywhere between 5-8 players at the same time, the database grew significantly from ~390mb to ~676mb. Over time that became 800mb at its peak, however as #10200 shows it has its performance drawbacks.

I therefore suggest to offer a choice between SQLite (for singleplayer/low pop servers) and either MySQL or PostgreSQL for medium to high population servers for performance reasons. This would additionally incentivize long-term servers which would benefit from a database not being wiped periodically for performance reasons.

**Comment by D3nnis3n** [comment]
On top, that database solution is a very slow one, causing lag. You got CM's support.

**Comment by elfl0rd** [comment]
ha, ha... i knew servers with db larger than 4gb, lol.
Totally agree.

**Comment by johnkslg** [comment]
Not sure it would reduce size to use another DB format. Perhaps there are some compression options we can add, or data usage optimization.

**Comment by D3nnis3n** [comment]
Mine has 5.3 GB. For me it's more about the lag. SQLite is not a very performant solution.

**Comment by DragonSupremacy** [comment]
Yeah, file size is secondary to be honest. My primary motivation for this suggestion however stems from one of the limitations of SQLite, namely the lack of concurrent write operations. It only allows for a single write operation to take place at the same time, meaning that if you buy a large quantity of stuff at a store at once, it'll immediately lag the server for several seconds (when playing alone) to much longer depending on how much of the various play actions are registered at the same time, which are all bottlenecked by SQLite's "one write operation at a time" limitation.

On the other hand, MySQL when used in conjunction with InnoDB allows the DB to be more multi-threaded and allowing concurrency to take place, e.g. registering the various play actions of multiple users playing at the same time a large purchase takes place will not be hindered by the large INSERT operation which buying large quantities of goods from a store is.

**Comment by D3nnis3n** [comment]
Exactly that.

Am Di., 18. Dez. 2018, 00:55 hat DragonSupremacy <notifications@github.com>
geschrieben:

**Comment by acmcpheron02** [comment]
I've also seen some performance issues when running a server, so I'm also in favor of a SQL based backend solution. Something worth considering as well is that it opens up some new modding potential with players being able to create custom data visualizations or market analysis. I don't have a lot of experience with NoSQL so maybe it's possible to do those already, but I'm assuming it's at least difficult when the database is embedded in the server client like that.
